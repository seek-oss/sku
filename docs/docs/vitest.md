# Vitest

[Vitest] is a testing framework that supports ESM out-of-the-box, integrates with the Vite ecosystem and has a similar API to Jest.
These features make it a great replacement for Jest in `sku` applications, especially given [Jest's current limitations with ESM].
Sku supports Vitest as an alternate to Jest since v15

**We recommend using vitest with the vite bundler and in an ESM project**, however these are not requirements.

## Migrating to Vitest

Migrating to Vitest should be fairly straightforward.
Before making any changes, it is highly recommend to read [the Vitest documentation](https://vitest.dev/guide/) to familiarise yourself with the API and features, as well as to better understand the differences between Jest and Vitest.

To enable Vitest in `sku`, first install the required dependencies:

```bash
pnpm add -D vitest @sku-lib/vitest
```

Then, configure [`testRunner`][test runner] in your `sku` config:

```typescript
// sku.config.ts
import type { SkuConfig } from 'sku';

export default {
  testRunner: 'vitest',
  ...
} satisfies SkuConfig;
```

`sku` will now invoke [the `vitest` CLI][Vitest CLI] instead of the `jest` CLI when running `sku test`.

#### Key differences between Vitest and Jest

Vitest has strong compatibility with the Jest API, however it still has differences that may affect your tests.
Differences that affect sku projects are listed here.
The full list is documented in the Vitest documentation: [Migrating from Jest to Vitest].

To automate most of the migration process, a codemod is available.
Note that additional changes may still be required after running this codemod.

```sh
pnpm dlx @sku-lib/codemod jest-to-vitest .
```

**Default Watch Mode**

`vitest` defaults to watch mode when running tests.
To run tests without watch mode you can use the `--run` flag:

```json
// package.json
{
  "scripts": {
    // ...
    "test": "sku test --run"
  }
}
```

Watch mode won't trigger in CI environments, so it's safe to omit the flag in your pipeline.

**Code Coverage Dependencies**

Vitest does not install code coverage dependencies by default.
To collect code coverage, pass the `--coverage` flag to the `sku test` command and install the `@vitest/coverage-v8` package:

```bash
pnpm add -D @vitest/coverage-v8
```

**Testing Library Matchers**

If your test setup file includes an import for `@testing-library/jest-dom`, you may need to change this to `@testing-library/jest-dom/vitest`:

```diff
// test-setup.ts
- import '@testing-library/jest-dom';
+ import '@testing-library/jest-dom/vitest';
```

**Globals Disabled**

Jest enables global APIs such as `it`, `describe`, `beforeAll`, etc., by default.
Vitest does not.
Sku adopts the Vitest defaults, meaning you will need to explicitly import these test functions.
This change results in a cleaner global namespace, better type safety, clearer dependencies and is more consistency with modern javascript practices.

```diff
+ import { describe, expect, it } from 'vitest';
```

Be aware that since globals are disabled, some common libraries like `testing-library` will not run auto DOM cleanup.
If using these libraries, you will need to add cleanup to your configured `setupTests` file.

```diff
import '@testing-library/jest-dom/vitest';

+ import { cleanup } from '@testing-library/react';
+ import { afterEach } from 'vitest';

+ afterEach(cleanup)
```

[Vitest]: https://vitest.dev/
[Jest's current limitations with ESM]: https://jestjs.io/docs/ecmascript-modules
[test runner]: ./docs/configuration.md#testrunner
[Vitest CLI]: https://vitest.dev/guide/cli.html
[codemod]: https://codemod.com/registry/jest-vitest
[Migrating from Jest to Vitest]: https://vitest.dev/guide/migration.html#jest
[Vitest globals]: https://vitest.dev/config/#globals
