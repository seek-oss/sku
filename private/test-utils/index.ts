export { ListExternalsWebpackPlugin } from './ListExternalsWebpackPlugin.ts';
export { dirContentsToObject } from './dirContentsToObject.ts';
export { makeStableHashes, makeStableViteHashes } from './skuConfig.ts';
export { getStoryPage, getTextContentFromFrameOrPage } from './storybook.ts';
export { getPort } from './getPort.ts';
