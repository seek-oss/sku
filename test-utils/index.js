// @ts-check
export { ListExternalsWebpackPlugin } from './ListExternalsWebpackPlugin.js';
export { getAppSnapshot } from './appSnapshot.js';
export { startAssetServer } from './assetServer.js';
export { dirContentsToObject } from './dirContentsToObject.js';
export { run, runSkuScriptInDir } from './process.ts';
export { makeStableHashes } from './skuConfig.js';
export { getStoryPage, getTextContentFromFrameOrPage } from './storybook.js';
export { waitForUrls } from './waitForUrls.js';
