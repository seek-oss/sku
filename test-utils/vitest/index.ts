export { getAppSnapshot } from './appSnapshot.ts';
