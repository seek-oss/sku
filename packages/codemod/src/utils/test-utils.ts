import { readFile } from 'node:fs/promises';
import path from 'node:path';
import { expect } from 'vitest';
import { isValidTransform, type Transform } from './types.js';

type FileInfo = {
  source: string;
  path: string;
};

type InlineTestProps = {
  transform: Transform;
  input: FileInfo;
  expectedOutput: string;
};

type TestProps = {
  dirName: string;
  transformName: string;
  fixtureName?: string;
  testOptions?: { extension?: string };
};

const runInlineTest = async ({
  transform,
  input,
  expectedOutput,
}: InlineTestProps) => {
  if (!isValidTransform(transform)) {
    throw new Error('Invalid transform function provided');
  }

  const output = transform(input.source);
  if (!output) {
    throw new Error('Transform failed');
  }

  expect(output.trim()).toEqual(expectedOutput.trim());

  return output;
};

export const runTest = async ({
  dirName,
  transformName,
  fixtureName,
  testOptions = {},
}: TestProps) => {
  const testFixtureName = fixtureName || transformName;

  const fixtureDir = path.join(dirName, '..', '__testfixtures__');
  const inputPath = path.join(
    fixtureDir,
    `${testFixtureName}.input.${testOptions.extension}`,
  );

  // Assumes transform is one level up from __tests__ directory
  const transformImport = import(path.join(dirName, '..', transformName));
  const sourceRead = readFile(inputPath, 'utf8');
  const expectedOutputRead = readFile(
    path.join(fixtureDir, `${testFixtureName}.output.${testOptions.extension}`),
    'utf8',
  );

  const [{ transform }, source, expectedOutput] = await Promise.all([
    transformImport,
    sourceRead,
    expectedOutputRead,
  ]);

  return runInlineTest({
    transform,
    input: {
      path: inputPath,
      source,
    },
    expectedOutput,
  });
};

export const runNoChangeTest = async ({
  dirName,
  transformName,
  fixtureName,
  testOptions = {},
}: TestProps) => {
  const inputPath = path.join(
    dirName,
    '..',
    '__testfixtures__',
    `${fixtureName}.unchanged.${testOptions.extension}`,
  );

  const sourceRead = readFile(inputPath, 'utf8');
  const transformImport = import(path.join(dirName, '..', transformName));

  const [{ transform }, source] = await Promise.all([
    transformImport,
    sourceRead,
  ]);
  const output = await transform(source);

  expect(output).toEqual(null);
};
