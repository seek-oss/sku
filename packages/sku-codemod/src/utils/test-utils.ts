import { readFile } from 'node:fs/promises';
import path from 'node:path';
import { expect } from 'vitest';

type FileInfo = {
  source: string;
  path: string;
};

const runInlineTest = async (
  transform: (source: string) => string | false,
  input: FileInfo,
  expectedOutput: string,
) => {
  const output = await transform(input.source);
  if (!output) {
    throw new Error('Transform failed');
  }
  const expectation = (o: string) =>
    expect(o.trim()).toEqual(expectedOutput.trim());
  expectation(output);
  return output;
};

export const runTest = async (
  dirName: string,
  transformName: string,
  testFilePrefix?: string,
  testOptions: Record<string, unknown> = {},
) => {
  let localTestFilePrefix = testFilePrefix;
  if (!testFilePrefix) {
    localTestFilePrefix = transformName;
  }

  // Assumes transform is one level up from __tests__ directory
  const transform = (await import(path.join(dirName, '..', transformName)))
    .transform;
  const fixtureDir = path.join(dirName, '..', '__testfixtures__');
  const inputPath = path.join(
    fixtureDir,
    `${localTestFilePrefix}.input.${testOptions.extension}`,
  );
  const source = await readFile(inputPath, 'utf8');
  const expectedOutput = await readFile(
    path.join(
      fixtureDir,
      `${localTestFilePrefix}.output.${testOptions.extension}`,
    ),
    'utf8',
  );
  return runInlineTest(
    transform,
    {
      path: inputPath,
      source,
    },
    expectedOutput,
  );
};

export const runNoChangeTest = async (
  dirName: string,
  transformName: string,
  testFilePrefix: string,
  testOptions: Record<string, unknown>,
) => {
  const transform = (await import(path.join(dirName, '..', transformName)))
    .transform;
  const inputPath = path.join(
    dirName,
    '..',
    '__testfixtures__',
    `${testFilePrefix}.unchanged.${testOptions.extension}`,
  );
  const source = await readFile(inputPath, 'utf8');
  const output = await transform(source);
  expect(output).toEqual(false);
};
