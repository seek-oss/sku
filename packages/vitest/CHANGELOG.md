# @sku-lib/vitest

## 0.2.1

### Patch Changes

- Fix `watch` mode ([#1393](https://github.com/seek-oss/sku/pull/1393))

## 0.2.0

### Minor Changes

- Disable Vitest globals ([#1387](https://github.com/seek-oss/sku/pull/1387))

  `vitest` is now a peer dependency. As such, `vitest` APIs must now be imported explicitly

  **EXAMPLE USAGE**:

  ```ts
  // example.test.ts
  import { describe, it, expect } from 'vitest';

  describe("test suite", () => {
    it("should be true" () => {
      expect(true).toBe(true);
    });
  });
  ```

### Patch Changes

- `deps`: Update `vite-plugin-cjs-interop` from `^2.2.0` to `^2.3.0` ([#1386](https://github.com/seek-oss/sku/pull/1386))

- Added missing `compilePackages` and `cjsInteropDependencies` support ([#1381](https://github.com/seek-oss/sku/pull/1381))

- Fixed a bug causing `sku test` to hang when `vitest` is executed in `run` mode ([#1387](https://github.com/seek-oss/sku/pull/1387))

## 0.1.1

### Patch Changes

- Experimental vite bundler now requires vite v7. ([#1346](https://github.com/seek-oss/sku/pull/1346))

  This is a breaking peer dependency change for projects that use the experimental vite bundler. Since this is an experimental feature, we are not bumping the major version.

- - `vitest`: Honour watch mode when running `sku test` ([#1325](https://github.com/seek-oss/sku/pull/1325))

    Previously it was not possible to run Vitest in watch mode via `sku test`. `sku test` should now behave similarly to [the `vitest` CLI][vitest cli], so `sku test`, `sku test watch` and `sku test -w/--watch` will all trigger watch mode.

    To run tests once and exit, you can execute `sku test run`, `sku test --run` or `sku test --no-watch` .

    [vitest cli]: https://vitest.dev/guide/cli

  - `deps` Bump peer dependencies to support React 19

## 0.1.0

### Minor Changes

- Initial release. ([#1273](https://github.com/seek-oss/sku/pull/1273))
