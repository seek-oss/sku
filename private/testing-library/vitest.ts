// Re-export vitest setup from cli-testing-library
import 'cli-testing-library/vitest';
