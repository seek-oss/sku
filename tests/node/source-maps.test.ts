import { describe, it, expect } from 'vitest';

import { dirContentsToObject } from '@sku-private/test-utils';
import {
  bundlers,
  type BundlerValues,
  scopeToFixture,
} from '@sku-private/testing-library';

const { sku, fixturePath } = scopeToFixture('source-maps');

describe('source-maps', () => {
  const args: BundlerValues<string[]> = {
    vite: ['--config', 'sku.config.vite.ts'],
    webpack: [],
  };

  describe.each(bundlers)('bundler %s', (bundler) => {
    describe.for([true, false])('sourceMapsProd %s', (sourceMapsProd) => {
      it('should generate the expected files', async () => {
        const build = await sku('build', args[bundler], {
          spawnOpts: {
            env: {
              ...process.env,
              SKU_SOURCE_MAPS_PROD: String(sourceMapsProd),
            },
          },
        });
        expect(await build.findByText('Sku build complete')).toBeInTheConsole();

        const files = await dirContentsToObject(fixturePath('./dist'));
        expect(files).toMatchSnapshot();
      });
    });
  });
});
