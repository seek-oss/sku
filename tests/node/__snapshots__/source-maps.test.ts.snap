// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`source-maps > bundler vite > sourceMapsProd false > should generate the expected files 1`] = `
{
  "index.html": SCRIPTS: [
    "/static/source-maps/vite-client-DcM1ITlA.js",
  ]
CSS: []
SOURCE HTML: <!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>
      hello-world
    </title>
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1"
    >
    <link
      rel="modulepreload"
      href="SCRIPTS[0]"
      crossorigin
    >
  </head>
  <body>
    <div id="app">
      <div>
        Hello source maps
      </div>
    </div>
    <script
      type="module"
      src="SCRIPTS[0]"
    >
    </script>
  </body>
</html>,
  "vite-client-DcM1ITlA.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
}
`;

exports[`source-maps > bundler vite > sourceMapsProd true > should generate the expected files 1`] = `
{
  "index.html": SCRIPTS: [
    "/static/source-maps/vite-client-DcM1ITlA.js",
  ]
CSS: []
SOURCE HTML: <!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>
      hello-world
    </title>
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1"
    >
    <link
      rel="modulepreload"
      href="SCRIPTS[0]"
      crossorigin
    >
  </head>
  <body>
    <div id="app">
      <div>
        Hello source maps
      </div>
    </div>
    <script
      type="module"
      src="SCRIPTS[0]"
    >
    </script>
  </body>
</html>,
  "vite-client-DcM1ITlA.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "vite-client-DcM1ITlA.js.map": "CONTENTS IGNORED IN SNAPSHOT TEST",
}
`;

exports[`source-maps > bundler webpack > sourceMapsProd false > should generate the expected files 1`] = `
{
  "2.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "index.html": SCRIPTS: [
    "/static/source-maps/runtime.js",
    "/static/source-maps/2.js",
    "/static/source-maps/main.js",
  ]
CSS: []
SOURCE HTML: <!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>
      hello-world
    </title>
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="SCRIPTS[0]"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="SCRIPTS[1]"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="SCRIPTS[2]"
    >
  </head>
  <body>
    <div id="app">
      <div>
        Hello source maps
      </div>
    </div>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS__"
      type="application/json"
    >
      []
    </script>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS___ext"
      type="application/json"
    >
      {"namedChunks":[]}
    </script>
    <script
      async
      data-chunk="main"
      src="SCRIPTS[0]"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="SCRIPTS[1]"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="SCRIPTS[2]"
    >
    </script>
  </body>
</html>,
  "main.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "runtime.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
}
`;

exports[`source-maps > bundler webpack > sourceMapsProd true > should generate the expected files 1`] = `
{
  "2.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "2.js.map": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "index.html": SCRIPTS: [
    "/static/source-maps/runtime.js",
    "/static/source-maps/2.js",
    "/static/source-maps/main.js",
  ]
CSS: []
SOURCE HTML: <!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>
      hello-world
    </title>
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="SCRIPTS[0]"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="SCRIPTS[1]"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="SCRIPTS[2]"
    >
  </head>
  <body>
    <div id="app">
      <div>
        Hello source maps
      </div>
    </div>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS__"
      type="application/json"
    >
      []
    </script>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS___ext"
      type="application/json"
    >
      {"namedChunks":[]}
    </script>
    <script
      async
      data-chunk="main"
      src="SCRIPTS[0]"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="SCRIPTS[1]"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="SCRIPTS[2]"
    >
    </script>
  </body>
</html>,
  "main.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "main.js.map": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "runtime.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "runtime.js.map": "CONTENTS IGNORED IN SNAPSHOT TEST",
}
`;
